File: day14_part1_2015.exs

defmodule ReindeerOlympics do
  def call do
    "input.txt"
    |> File.read!()
    |> String.split("\n", trim: true)
    |> Enum.map(&parse_reindeer/1)
    |> Enum.map(&calculate_distance(&1, 2503))
    |> Enum.max()
  end

  defp parse_reindeer(line) do
    [speed, fly_time, rest_time] =
      Regex.scan(~r/\d+/, line)
      |> List.flatten()
      |> Enum.map(&String.to_integer/1)

    {speed, fly_time, rest_time}
  end

  defp calculate_distance({speed, fly_time, rest_time}, total_time) do
    cycle_time = fly_time + rest_time
    full_cycles = div(total_time, cycle_time)
    remaining_time = rem(total_time, cycle_time)
    distance_per_cycle = speed * fly_time
    total_distance = full_cycles * distance_per_cycle
    remaining_distance = speed * min(remaining_time, fly_time)

    total_distance + remaining_distance
  end
end

File: prolog/day14_part1_2015.pl
:- initialization(main).

main :-
    read_reindeer_data('input.txt', Reindeer),
    race(Reindeer, 2503, Distances),
    max_list(Distances, WinningDistance),
    format('Winning distance: ~w~n', [WinningDistance]),
    halt.

read_reindeer_data(File, Reindeer) :-
    open(File, read, Stream),
    read_lines(Stream, Reindeer),
    close(Stream).

read_lines(Stream, []) :-
    at_end_of_stream(Stream), !.
read_lines(Stream, [R|Rs]) :-
    \+ at_end_of_stream(Stream),
    read_line_to_string(Stream, Line),
    parse_reindeer(Line, R),
    read_lines(Stream, Rs).

parse_reindeer(Line, reindeer(Name, Speed, FlyTime, RestTime)) :-
    split_string(Line, " ", " ", Parts),
    nth0(0, Parts, Name),
    nth0(3, Parts, SpeedStr),
    nth0(6, Parts, FlyTimeStr),
    nth0(13, Parts, RestTimeStr),
    atom_number(SpeedStr, Speed),
    atom_number(FlyTimeStr, FlyTime),
    atom_number(RestTimeStr, RestTime).

race(Reindeer, Time, Distances) :-
    findall(Distance, (member(reindeer(_, Speed, FlyTime, RestTime), Reindeer), 
                      total_distance(Speed, FlyTime, RestTime, Time, Distance)), Distances).

total_distance(Speed, FlyTime, RestTime, Time, Distance) :-
    CycleTime is FlyTime + RestTime,
    FullCycles is Time // CycleTime,
    RemainingTime is Time mod CycleTime,
    DistanceCovered is FullCycles * Speed * FlyTime + min(RemainingTime, FlyTime) * Speed,
    Distance is DistanceCovered.